hermes-consumers/src/main/java/pl/allegro/tech/hermes/consumers/consumer/SendFutureProvider.java

```diff
@@ -0,0 +1,92 @@
+package pl.allegro.tech.hermes.consumers.consumer;
+
+import pl.allegro.tech.hermes.api.Subscription;
+import pl.allegro.tech.hermes.consumers.consumer.rate.ConsumerRateLimiter;
+import pl.allegro.tech.hermes.consumers.consumer.sender.MessageSendingResult;
+import pl.allegro.tech.hermes.consumers.consumer.sender.timeout.FutureAsyncTimeout;
+
+import java.time.Duration;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+// provides futures which are rate limited and completed exceptionally after given timeout
+public class SendFutureProvider {
```

Comment:
In my opinion, this name is too generic. The main responsibility of this class is to apply the rate limiter to outgoing requests. So maybe `RateLimitingMessageSender` (or something along those lines) would be more descriptive. The only problem I see is that we also deal with error mapping here - could it be done elsewhere?

Reply:
OK. I like your suggestion - `ResillientMessageSender`.